Port search commands (`/`) to newshell

DMaroo opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

This PR ports the search commands (`/`) to Rizin newshell.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Once the porting is done, the build and CI should be green. Maybe I'll add tests for any new features that get added during the porting.

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->
Closes #1586.
Tracked by #1342.

DMaroo committed:
Add `rz_str_extract_printable` util function

DMaroo committed:
Start porting search commands (`/`) to newshell

DMaroo committed:
Use `RZ_STR_DUP` macro instead of `rz_str_newf`

DMaroo committed:
Rename `eprintf` to `RZ_LOG_ERROR`

DMaroo committed:
Assign correct values to `param` members

DMaroo committed:
Add `UPDATE_LASTSEARCH` macro

DMaroo committed:
Open and end the JSON object appropriately

DMaroo committed:
Fixed `search_kw` null handling

XVilka commented on librz/util/str.c:
```diff
@@ -2203,6 +2203,29 @@ RZ_API void rz_str_filter(char *str, int len) {
 }
 
+/**
+ * \brief Exttract all printable characters in \p str, and return a string of maximum length \p len
```
A typo

XVilka commented on librz/util/str.c:
```diff
@@ -2203,6 +2203,29 @@ RZ_API void rz_str_filter(char *str, int len) {
+ * \param len The maximum length of the resulting string
+ * \return char* Extracted printable string
+ */
+RZ_API char *rz_str_extract_printable(const char *str, int len) {
```
Add ownership mark for the return, `RZ_NONULL` to `str`

XVilka commented on librz/util/str.c:
```diff
@@ -2203,6 +2203,29 @@ RZ_API void rz_str_filter(char *str, int len) {
+RZ_API char *rz_str_extract_printable(const char *str, int len) {
+	rz_return_val_if_fail(str && len >= 0, NULL);
+	char *r, *res = malloc(len + 1);
+	int i;
```
Move `int i;` after error check

DMaroo committed:
Rearrange functions and remove "currently" unused functions

  * Also replaced eprintf with RZ_LOG_ERROR
  * Build and unit tests pass now

DMaroo committed:
Change according to review

DMaroo committed:
Add `RZ_NONNULL` decoration for function arguments in `str.c`

DMaroo committed:
Format according to clang-format

DMaroo committed:
Fix segfault and remove unnecessary RZ_LOG_ERROR

DMaroo committed:
Add details for `cx` command regarding ignoring bytes while comparing

DMaroo committed:
Port `/x` search commands; Fix JSON printing

DMaroo committed:
Implement command handler for `/a`

  * `/aj` is unimplemented yet

DMaroo committed:
Add `RzSearchParams` struct

  * This will allow for a generic search function, since all search
    parameters can be represented using the `RzSearchParams` struct
  * Also, the struct contains search parameters, so any search function
    (whose job is to "search") would never need to modify any of these
    params. This allows for some sort of contraint/invariance and will
    help in making sure that the search functions which get implemented
    are correct.
  * Also added `search.minlength` and `search.maxlength` in `cconfig.c`

ret2libc commented on librz/util/str.c:
```diff
@@ -2201,6 +2201,29 @@ RZ_API void rz_str_filter(char *str, int len) {
+		return NULL;
+	}
+	int i;
+	for (i = 0, r = res; i < len; str++, i++) {
```
Done this way it seems that len is the maximum length of `str` and not of the resulting string.

ret2libc commented on librz/core/cmd_descs/cmd_cmp.yaml:
```diff
@@ -157,6 +157,11 @@ commands:
     args:
       - name: hexpair
         type: RZ_CMD_ARG_TYPE_STRING
+    details:
+      - name: Ignoring bytes
+        entries:
+          - text: "Ignore specific bytes while comparing by using '.'"
+            comment: "cx 1ee7..0f....d34d"
```
I suggest to use the "regular" `Examples` section.
Also, the command `cx` should co in `text`, the argument `1ee7..0f....d34d` in `arg_str` and `"Ignore specific bytes while comparing by using '.'"` in `comment`.

ret2libc commented on librz/core/cmd/cmd_print.c:
```diff
@@ -844,6 +844,36 @@ static const ut32 colormap[256] = {
 	0xffffff,
 };
 
+static void incAlphaBuffer(ut8 *buf, int bufsz) {
```
what are these?

DMaroo replied:
These are some functions that I needed to keep such that the build passes, I will refactor them in the end.

ret2libc commented on librz/core/cmd/cmd_search.c:
```diff
@@ -438,13 +62,13 @@ static int _cb_hit(RzSearchKeyword *kw, void *user, ut64 addr) {
 			rz_io_read_at(core->io, addr - prectx, (ut8 *)buf, len + (ctx * 2));
-			pre = getstring(buf, prectx);
-			pos = getstring(buf + prectx + len, ctx);
+			pre = rz_str_extract_printable(buf, prectx);
```
add a unit test for the new rz_str_extract_printable API.

ret2libc commented on librz/core/cmd/cmd_search.c:
```diff
@@ -529,3338 +153,363 @@ static int _cb_hit(RzSearchKeyword *kw, void *user, ut64 addr) {
 	if (rz_cons_singleton()->columns < 50) {
-		eprintf("\r[  ]  0x%08" PFMT64x "  hits = %d   \r%s",
-			at, n, (c % 2) ? "[ #]" : "[# ]");
+		rz_cons_printf("\r[  ]  0x%08" PFMT64x "  hits = %d   \r%s",
```
i'd use RZ_LOG_INFO because this is not the real output of the command, it's just some update on the status.

ret2libc reviewed (commented):
I think the RzSearchParameters split make sense, but I'd have a new `_new` API that accepts a `RzSearchParameters` structure to initialize the search. I think starting writing unit tests will definitely help to find out what is the best way to deal with this.

I'd see something like the following:
```C
RzSearchParams params = {
.mode = RZ_SEARCH_STRING,
.string_min = 3,
.string_max = 10,
};
RzSearch *s = rz_search_new(&params);
```

note that some of the parameters may be meaningful only for a specific "search mode". In such cases it may make sense to have RzSearchParameters composed of a union for the data, selected by the `search_mode` field.

DMaroo committed:
Add new and free functions for `RzSearchParams`

  * Rename `RzSearchParameters` to `RzSearchParams`
  * Rename struct member names
  * Add `rz_search_params_new`, `rz_core_search_params_new` and
    `rz_search_params_free`
  * Rename `rz_search_set_mode` to `rz_search_params_set_mode`

ret2libc commented on librz/include/rz_search.h:
```diff
@@ -51,39 +51,62 @@ typedef struct rz_search_hit_t {
+	RzSearchParams *params;
 } RzSearch;
 
+typedef struct rz_core_t RzCore;
```
This is not good here. There should be no reference to core

DMaroo replied:
Yeah, it's no longer necessary. I'll remove it.

DMaroo committed:
Remove `RzCore` from `rz_search.h`

DMaroo committed:
Fix the logic in `rz_str_extract_printable`

DMaroo committed:
Use `RZ_LOG_INFO` instead of `rz_cons_printf`

DMaroo committed:
Change details for `cx`

DMaroo committed:
Remove `rz_core_search_params_new`; Use `rz_search_params_new` instead

DMaroo committed:
Add unit test for `rz_str_extract_printable`

DMaroo committed:
Replace `rz_search_set_distance` and `rz_search_kw_add` with their `RzSearchParams` counterpart

DMaroo committed:
Initialize search parameters in `rz_core_init`

DMaroo committed:
Add ownership annotation for `rz_search_new`

DMaroo committed:
Format according to clang-format

DMaroo committed:
Replace `rz_search_string_prepare_backward` with `RzSearchParams` variant

DMaroo committed:
Use `RzSearchParam` callback in `rz_search_find` if it exists

DMaroo committed:
Removed unused duplicate function `rz_search_update_i`

DMaroo committed:
Replace `eprintf`; Fix code typos

DMaroo committed:
Remove unused functions `rz_search_find`, `listcb`

DMaroo committed:
Fix merge conflict remnants

XVilka commented:
It needs rebase. If you finish some particular command or few of them - probably better to focus on that, so the porting is incremental. Otherwise, we have to postpone this for 0.6.0

XVilka commented:
@DMaroo do you plan to continue working on this? Otherwise it probably better to close this one and start from scratch.

DMaroo commented:
@XVilka I have been very low on time lately. So I don't expect any progress on this for a while. Closing it seems reasonable. I'll open a new PR whenever I get done with the porting.

DMaroo commented:
Changes have become stale and there are too many conflicts. Closing this PR for now. ETA on this is undecided.

ret2libc commented:
I'm going to work on this as discussed on mattermost. I will start with `/R` and then expand from there. I will just try to convert the commands for now and leave the RzSearch refactoring to later.
